osquery/events/linux/bpf/processcontextfactory.cpp

```diff
@@ -0,0 +1,279 @@
+  // such as a kworker instance
+  if (output.binary_path.empty() != output.argv.empty()) {
+    return false;
+  }
+
+  if (!fs.readLinkAt(output.cwd, process_root.get(), "cwd")) {
+    return false;
+  }
+
+  if (!getParentPidFromStatFile(
+          fs, output.parent_process_id, process_stat.get())) {
+    return false;
+  }
+
+  process_context = std::move(output);
+  return true;
+}
+
+bool ProcessContextFactory::captureAllProcesses(
+    IFilesystem& fs, ProcessContextMap& process_map) {
+  process_map = {};
+
+  tob::utils::UniqueFd process_root;
+  if (!fs.open(process_root, kProcFsRoot, O_DIRECTORY)) {
+    return false;
+  }
+
+  ProcessContextMap output;
+
+  // clang-format off
+  auto succeeded = fs.enumFiles(
+    process_root.get(),
+
+    [&](const std::string &name, bool directory) {
+      if (!directory) {
+        return;
+      }
+
+      char* null_terminator{nullptr};
+      auto pid = static_cast<pid_t>(std::strtoull(name.c_str(), &null_terminator, 10));
+      if (null_terminator == nullptr || *null_terminator != '\0') {
+        return;
+      }
+
+      ProcessContext process_context = {};
+      if (captureSingleProcess(fs, process_context, pid)) {
+        output.insert({pid, std::move(process_context)});
+      } else {
+      }
+    }
+  );
+  // clang-format on
+
+  process_map = std::move(output);
+  return succeeded;
+}
+
+bool ProcessContextFactory::getArgvFromCmdlineFile(
+    IFilesystem& fs, std::vector<std::string>& argv, int fd) {
+  argv = {};
+
+  std::vector<char> buffer;
+  if (!fs.read(buffer, fd, kMaxFileSize)) {
+    return false;
+  }
+
+  std::vector<std::string> output;
+  std::size_t start = 0U;
+
+  for (;;) {
+    std::size_t end;
+    for (end = start; end < buffer.size() && buffer.at(end) != 0; ++end)
+      ;
+
+    auto argument_size = end - start;
+    if (argument_size == 0) {
+      break;
+    }
+
+    auto argument = std::string(buffer.data() + start, argument_size);
+    output.push_back(std::move(argument));
+
+    start = end + 1;
+    if (start >= buffer.size()) {
+      break;
+    }
+  }
+
+  argv = std::move(output);
+  return true;
+}
+
+bool ProcessContextFactory::getParentPidFromStatFile(IFilesystem& fs,
+                                                     pid_t& parent_pid,
+                                                     int fd) {
+  parent_pid = -1;
+
+  std::vector<char> buffer;
+  if (!fs.read(buffer, fd, kMaxFileSize)) {
+    return false;
+  }
+
+  auto parent_pid_r_it = std::find(buffer.rbegin(), buffer.rend(), ')');
+  if (parent_pid_r_it == buffer.rend()) {
+    return false;
+  }
+
+  auto parent_pid_it = std::next(parent_pid_r_it.base(), 4);
+  if (parent_pid_it >= buffer.end()) {
+    return false;
+  }
+
+  auto parent_pid_index = parent_pid_it - buffer.begin();
+  auto parent_pid_ptr = &buffer[parent_pid_index];
+
+  char* space_separator_ptr{nullptr};
+  auto output_pid = std::strtoull(parent_pid_ptr, &space_separator_ptr, 10);
```

Comment:
Ditto `tryTo`.

Reply:
The std::strtoull in this case is required because we are relying on finding the space separator; tryTo does not support this and will return an error